File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.js
/**
 * @param {string} s
 * @return {number}
 */
var longestPalindromeSubseq = function(s) {
    const sRev = s.split("").reverse().join();
    return longestCommonSubsequence(s, sRev);
};

var longestCommonSubsequence = function (text1, text2) {
    return lcsSpaceOptimized(text1, text2);
};

function lcsSpaceOptimized(text1, text2) {
    const textLen1 = text1.length, textLen2 = text2.length;
    let prev = new Array(textLen2 + 1), curr = new Array(textLen2 + 1).fill(0);
    for (let j = 0; j <= textLen2; j++) {
        prev[j] = 0;
    }

    for (let i = 1; i <= textLen1; i++) {
        for (let j = 1; j <= textLen2; j++) {
            if (text1[i - 1] === text2[j - 1]) {
                // both match
                curr[j] = 1 + prev[j - 1];
            } else {
                // doesn't match
                curr[j] = Math.max(prev[j], curr[j - 1]);
            }
        }
        prev = [...curr]; ///// KEEP IN MIND TO DO LIKE THIS INSTEAD OF prev = curr
    }

    return prev[textLen2];
}

File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.cpp
class Solution {
public:
    int longestPalindromeSubseq(string s) {
        string revS = s;
        reverse(revS.begin(), revS.end());
        int m = s.size();
        vector<vector<int>> dp(m+1, vector<int>(m+1));
        for(int i=0; i<=m; i++) {
            for(int j=0; j<=m; j++) {
                if(i == 0 || j == 0) {
                    dp[i][j] = 0;
                } else if(s[i-1] == revS[j-1]) {
                    dp[i][j] = 1 + dp[i-1][j-1];
                } else {
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
                }
            }
        }
        return dp[m][m];
    }
};